### PR TITLE
Extract type mapping postprocessing to an external visitor

### PR DESCRIPTION
**Pull request by roji**
Part of API review (#33220)

This extracts RelationalInferredTypeMappingApplier out of the translation visitor and makes it its own, independent postprocessing visitor. Note that I ended up not introducing a parameter object, but just passing in the general postprocessing dependencies that RelationalQueryTranslationPostprocessor has.

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs [hunk 1]
```diff
@@ -36,31 +37,45 @@ public class RelationalQueryTranslationPostprocessor : QueryTranslationPostproce
...
 
-        var query5 = new SqlExpressionSimplifyingExpressionVisitor(RelationalDependencies.SqlExpressionFactory, _useRelationalNulls)
-            .Visit(query4);
-        var query6 = new RelationalValueConverterCompensatingExpressionVisitor(RelationalDependencies.SqlExpressionFactory).Visit(query5);
+        var query6 = new SqlExpressionSimplifyingExpressionVisitor(RelationalDependencies.SqlExpressionFactory, _useRelationalNulls)
+            .Visit(query5);
+        var query7 = new RelationalValueConverterCompensatingExpressionVisitor(RelationalDependencies.SqlExpressionFactory).Visit(query6);
 
-        return query6;
+        return query7;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:66]: I see a weakness in this naming scheme

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -1328,11 +1323,11 @@ protected override ShapedQueryExpression TranslateUnion(ShapedQueryExpression so
     ///     Inferred type mappings for queryable constants/parameters collected during translation. These will be applied to the appropriate
     ///     nodes in the tree.
     /// </param>
+    [Obsolete("Override RelationalQueryTranslationPostprocessor.ProcessTypeMappings() instead.", error: true)]
     protected virtual Expression ApplyInferredTypeMappings(
         Expression expression,
         IReadOnlyDictionary<(string, string), RelationalTypeMapping?> inferredTypeMappings)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1329]: You'd need to revert this to `IReadOnlyDictionary<(TableExpressionBase, string), RelationalTypeMapping?>` for the obsolete attribute to make sense

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteTypeMappingPostprocessor.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    protected override Expression VisitExtension(Expression expression)
+    {
+        switch (expression)
+        {
+            case JsonEachExpression jsonEachExpression
+                when TryGetInferredTypeMapping(jsonEachExpression.Alias, "value", out var typeMapping):
```
Comment (AndriySvyryd) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteTypeMappingPostprocessor.cs:49]: Should these literals be extracted?

File: src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:66]: Yeah.. The main goal here was to allow examining the different stages of postprocessing when debugging, i.e. not overwrite the same `query` variable with each successive pipeline step. Adding steps seems rare enough that it's OK - but I'm open to other ideas...

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteTypeMappingPostprocessor.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteTypeMappingPostprocessor.cs:49]: Yeah, a cleanup around this is probably in order.

File: src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:66]: You could name them according to the last step: `baseProcessedQuery`, `typeMappedQuery`, `projectionAppliedQuery`, etc.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:66]: Or go the QBASIC route: `query10`, `query20`, `query30` ...

Comment (roji) [reply src/EFCore.Relational/Query/RelationalQueryTranslationPostprocessor.cs:66]: I'm worried about scalability, with future visitors coming to visit we may need to go into the hundreds/thousands here.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1329]: Re-introducing TableExpressionBase seems like a bit much... Given that the visitor base class (RelationalInferredTypeMappingApplier) is also being obsoleted, I'll just update the message there and remove this altogether - it should be more than enough to point provider writers in the right direction.

**Pull request merged**